### PR TITLE
Improve DateMonthYearWidget component

### PR DESCRIPTION
**Pull request by ariya**
* migrate to TypeScript, add type annotations
* include props definition
* support Storybook
* add a basic unit test

Verify manually by creating a dashboard with a parameter:
1. Browse data, Sample Database, Products
2. Save as a question, add to a dashboard
3. Edit dashboard, Add a filter, choose `Time` and `Month and Year`
4. Column to filter on `Created At`, Done

Ensure that everything works exactly the same **before** and **after** this PR.

image

To view this component in Storybook: `yarn storybook` and search for `DateMonthYearWidget`:

image

### CODE REVIEW

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    const { container } = render(
+      <DateMonthYearWidget
+        value={"2021-07"}
+        setValue={jest.fn()}
+        onClose={jest.fn()}
+      ></DateMonthYearWidget>,
+    );
+
+    expect(screen.getByText("January")).toBeDefined();
+    expect(screen.getByText("December")).toBeDefined();
```
Comment (kulyk) [inline frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx:17]: Please use the [`toBeInTheDocument`](https://github.com/testing-library/jest-dom#tobeinthedocument) or [`toBeVisible`](https://github.com/testing-library/jest-dom#tobevisible) assertions for UI

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    // 07 = July and year 2021
+    expect(screen.getByRole("button")).toHaveTextContent("2021");
+    expect(container.querySelector('div[class*="bg-brand"]')).toHaveTextContent(
+      "July",
+    );
```
Comment (kulyk) [inline frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx:23]: Could be a good opportunity to use `aria-selected` here

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.tsx [hunk 3]
```diff
@@ -8,9 +7,25 @@ import YearPicker from "metabase/components/YearPicker";
...
+type State = {
+  month: number | null;
+  year: number;
+};
+
+class DateMonthYearWidget extends React.Component<Props> {
+  state: State = {
+    month: null,
+    year: moment().year(),
+  };
```
Comment (kulyk) [inline frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.tsx:25]: I guess the preferred way to define the `state` type is to do `React.Component<Props, State>`

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx (hunk 1 above)
Comment (ariya) [reply frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx]: Good idea, I've changed it accordingly.

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.tsx (hunk 3 above)
Comment (ariya) [reply frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.tsx]: Done!

File: frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx (hunk 2 above)
Comment (ariya) [reply frontend/src/metabase/components/DateMonthYearWidget/DateMonthYearWidget.unit.spec.tsx]: Spot on, I've added `aria-selected`.

**Comment by kulyk** [comment]
@ariya thanks for addressing all the comments 🙏 Happy to approve, one thing I noticed is that you accidentally committed `calva/output-window/output.calva-repl` file

**Comment by ariya** [comment]
> one thing I noticed is that you accidentally committed `calva/output-window/output.calva-repl` file

Oops, sorry about that. It should be fixed now.

**Review by kulyk — APPROVED** [review]
✨

**Review by daltojohnso — APPROVED** [review]

**Pull request merged**